File: c/testsrc/ptypestest/test_ptypes.cc
// Copyright 2020 Jacob Glueck
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <type_traits>

#include "tako/ptypes_runtime.hh"
#include "tako/helpers.hh"

#include "catch2/catch.hpp"
#include "tako/ptypes.hh"
#include "test_types/ptypes_test_types.hh"

using namespace test_types;
using namespace tako;

template <typename T>
static void simple_string_test() {
    std::string_view input = "hello world";
    auto str = make_ptype_string<T>(input).value();
    CHECK(make_string_view(str) == input);
}
TEST_CASE("ptype_string") {
    simple_string_test<ptypes::StringL8>();
    simple_string_test<ptypes::StringL16>();
    simple_string_test<ptypes::StringL32>();
}

TEST_CASE("ptype_maybe_num") {
    ptypes_test_types::Optional some_num {.maybe_num = ptypes::Lu32 {.value = 42}};
    ptypes_test_types::Optional none_num = {.maybe_num = ptypes::Empty {}};

    some_num.maybe_num.match(
        [&](const ptypes::Empty&) { CHECK(false); },
        [&](const ptypes::Lu32& some) { CHECK(some.value == 42); }
    );

   none_num.maybe_num.match(
        [&](const ptypes::Empty&) { },
        [&](const ptypes::Lu32&) { CHECK(false); }
    );

    auto some_num_data = some_num.serialize();
    auto some_num_parsed = expect_parse<ptypes_test_types::OptionalView>(some_num_data);

    auto none_num_data = none_num.serialize();
    auto none_num_parsed = expect_parse<ptypes_test_types::OptionalView>(none_num_data);

    some_num_parsed.maybe_num().match(
        [&](const ptypes::EmptyView&) { CHECK(false); },
        [&](const ptypes::Lu32View& some) { CHECK(some.value() == 42); }
    );

    none_num_parsed.maybe_num().match(
        [&](const ptypes::EmptyView&) { },
        [&](const ptypes::Lu32View&) { CHECK(false); }
    );
}

File: java/testsrc/tako/TestPtypes.java
// Copyright 2020 Jacob Glueck
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package tako;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import tako.ParseException;
import static tako.PtypesRuntime.*;
import takogen.tako.Ptypes;
import takogen.test_types.PtypesTestTypes;
import static tako.Helpers.*;

public class TestPtypes {
    @Test
    public void ptypeString() {
        var start = "hello world";
        var x = makePtypeStringL16(start);
        Assertions.assertEquals(start, makeString(x));
    }

    @Test
    public void maybeNum() throws ParseException {
        var someNum = new PtypesTestTypes.Optional();
        someNum.maybeNum().set(Ptypes.Lu32.class).setValue(42);
        var noneNum = new PtypesTestTypes.Optional();
        noneNum.maybeNum().set(Ptypes.Empty.class);

        someNum.maybeNum().accept(new PtypesTestTypes.MaybeNum.VoidVisitor() {
            public void visit(Ptypes.Lu32 x) {
                Assertions.assertEquals(x.value(), 42);
            }
            public void visit(Ptypes.Empty x) {
                Assertions.fail("Should not be empty");
            }
        });
        noneNum.maybeNum().accept(new PtypesTestTypes.MaybeNum.VoidVisitor() {
            public void visit(Ptypes.Lu32 x) {
                Assertions.fail("Should not be an Lu32");
            }
            public void visit(Ptypes.Empty x) {}
        });

        var someNumView = new PtypesTestTypes.OptionalView();
        someNumView.parse(someNum.serialize(), 0);
        var noneNumView = new PtypesTestTypes.OptionalView();
        noneNumView.parse(noneNum.serialize(), 0);

        someNumView.maybeNum().accept(new PtypesTestTypes.MaybeNumView.VoidVisitor() {
            public void visit(Ptypes.Lu32View x) {
                Assertions.assertEquals(x.value(), 42);
            }
            public void visit(Ptypes.EmptyView x) {
                Assertions.fail("Should not be empty");
            }
        });
        noneNumView.maybeNum().accept(new PtypesTestTypes.MaybeNumView.VoidVisitor() {
            public void visit(Ptypes.Lu32View x) {
                Assertions.fail("Should not be an Lu32");
            }
            public void visit(Ptypes.EmptyView x) {}
        });
    }
}
